Add init on behalf, PDA seeds and event structure changes

skr4n opened the pull request:
Please read the commit messages for more info.

skr4n committed:
fix: Expand PDA seeds to include all swap params, emit all params

While having just initiator and secret_hash is safe for the current
design, a soon to be added feature known as initiate on behalf
requires initiates without the initiator's signature.
This would allow for a scenario known as "frontrunning"" where
malicious validators craft transactions with swap params resulting in
the same PDAs, causing a denial of service.
Expanding the seeds makes the already expensive frontrunning scenario,
much more expensive, as the attacker would have to set the same
params, as well as lose the ability to refund their funds.

The previous Redeem, Refund, InstantRefund events used to only emit
initiator and secret hash as the PDAs guaranteed no two swaps exist at
the same time with these two common.
Since that's changing, the event structures have been updated to emit
all swap params, for off-chain verification.

skr4n committed:
refactor: Rename `expires_in_slots` to `timelock`

skr4n committed:
refactor: Rename `sponsor` to `rent_sponsor`

skr4n committed:
feat: Separate initiator into `funder` and `refundee`

The initiator in an atomic swap has two roles, one of depositing funds
and the other of obtaining refunds.
For cases where these two need to be different, e.g an initiator
wants to participate but cannot provide their signature,
they can do so through a party referred to as the "funder" who
deposits funds on the initiator's behalf.
And if the swap fails, the initiator aka refundee gets back the funds.

KatrixReloaded commented on programs/solana-spl-swaps/src/lib.rs:
```diff
@@ -323,9 +367,9 @@ pub struct Redeem<'info> {
     #[account(mut, token::authority = swap_data.redeemer)]
```
Can the caller set a different redeemer value in `ctx.swap_data.redeemer`? Just to be clear

skr4n replied:
This value reflects the exact redeemer value that's been passed to `initiate()` via its function params.
Since the PDA has been verified by this instruction above, this means the value of redeemer belongs to a successful initiate instruction.

KatrixReloaded replied:
Ok

skr4n commented on programs/solana-spl-swaps/src/lib.rs:
```diff
@@ -306,12 +344,18 @@ pub struct Redeem<'info> {
+            &swap_data.timelock.to_le_bytes(),
+        ],
+        bump = swap_data.bump,
+        close = rent_sponsor,
```
The `close` attribute here is responsible for deleting the PDA, and transferring the rent to the `rent_sponsor`

KatrixReloaded replied:
In the back-end, are you handling a scenario where someone might try to reuse a secret hash or is it allowed?

skr4n replied:
The solana program doesn't stop secret hash from being reused.
The watcher only verifies whether the on-chain params correspond with the database, as such it has no sense of whether a secret hash is "duplicate".
The secret hashes are determined during create order phase, generated by the FE I think, @Revantark can tell whether a potential duplicate of a secret hash is allowed there.

skr4n committed:
fix: Panic if `timelock` causes an integer overflow

Solana doesn't enable overflow protection for math operations by default.
Thus use safe additions and panic when an overflow happens.
A panic on the solana runtime causes the transaction to revert.

skr4n committed:
Merge branch 'fix/verify-mint' into feat/init-on-behalf

harshasingamshetty1 committed:
Made the Indentity PDA virtual and other refactors

skr4n committed:
refactor: minor styling and change in seed order

skr4n committed:
feat: Emit funder address in initiate events

In cases where the funder is a flagged address, allow solvers to
verify the funder by including in the `Initiated` event.

skr4n commented:
few things never tend to make it.
